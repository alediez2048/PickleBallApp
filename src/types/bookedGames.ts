// BookedGame type for booked games feature
// All fields must match the structure used in Supabase and the app

export interface BookedGame {
  id: string;
  game_id: string;
  date: string;
  time: string;
  court_name: string;
  location_id: string | null;
  skill_rating: number;
  price: number;
  user_id: string;
  user_info:{} | any
  status: "upcoming" | "completed" | "cancelled";
}
