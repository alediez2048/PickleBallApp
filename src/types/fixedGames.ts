import type { SkillLevel } from "@/types/skillLevel";

// Types for Fixed Games based on the SQL schema
export type DayOfWeek =
  | 'Sunday'
  | 'Monday'
  | 'Tuesday'
  | 'Wednesday'
  | 'Thursday'
  | 'Friday'
  | 'Saturday';

export type FixedGameStatus = 'active' | 'inactive';

export interface FixedGame {
  id: string;
  title: string;
  description?: string;
  day_of_week: DayOfWeek;
  start_time: string; // 'HH:MM:SS' format
  duration_minutes: number;
  location_id: string;
  location?: any; // Add location for joined data
  host: any; // JSON object, can be typed further if needed
  max_players: number;
  skill_level: SkillLevel;
  price: number;
  image_url?: string;
  status: FixedGameStatus;
  created_at: string;
  updated_at: string;
}

export type FixedGameInput = Omit<FixedGame, 'id' | 'created_at' | 'updated_at'>;
