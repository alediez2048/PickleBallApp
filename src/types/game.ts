export const SkillLevel = {
  Beginner: 'Beginner',
  Intermediate: 'Intermediate',
  Advanced: 'Advanced',
  Open: 'Open',
} as const;

export type SkillLevel = typeof SkillLevel[keyof typeof SkillLevel];

export const GameStatus = {
  Upcoming: 'Upcoming',
  InProgress: 'InProgress',
  Completed: 'Completed',
  Cancelled: 'Cancelled',
} as const;

export type GameStatus = typeof GameStatus[keyof typeof GameStatus];

export interface Location {
  id: string;
  name: string;
  address: string;
  city: string;
  state: string;
  zipCode: string;
  coordinates: {
    latitude: number;
    longitude: number;
  };
  imageUrl?: string;
}

export interface User {
  id: string;
  name: string;
  email: string;
  skillLevel: SkillLevel;
  rating?: number;
}

export interface Game {
  id: string;
  title: string;
  description: string;
  startTime: string;
  endTime: string;
  location: Location;
  host: User;
  players: User[];
  maxPlayers: number;
  skillLevel: SkillLevel;
  price: number;
  imageUrl?: string;
  status: GameStatus;
  createdAt: string;
  updatedAt: string;
}

export interface GameFilters {
  skillLevel?: SkillLevel;
  date?: string;
  location?: string;
  status?: GameStatus;
} 