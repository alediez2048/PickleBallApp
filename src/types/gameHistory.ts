// Game history type for PickleBallApp

export interface GameHistory {
  id: string;
  date: string;
  result: "win" | "loss";
  score: string;
  opponent: string;
}
