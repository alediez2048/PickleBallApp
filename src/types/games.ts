import type { Location } from "./locations";

export const SkillLevel = {
  Beginner: "Beginner",
  Intermediate: "Intermediate",
  Advanced: "Advanced",
  Open: "Open",
} as const;

export type SkillLevel = (typeof SkillLevel)[keyof typeof SkillLevel];

export const GameStatus = {
  Upcoming: "upcoming",
  Completed: "completed",
  Cancelled: "cancelled",
} as const;

export type GameStatus = (typeof GameStatus)[keyof typeof GameStatus];

export interface User {
  id: string;
  name: string;
  email: string;
  skill_level: SkillLevel;
  phone_number?: string;
  rating?: number;
  profile_image?: string | null;
}

export const GAME_CONSTANTS = {
  MAX_PLAYERS: 8,
  MIN_PLAYERS: 2,
  DEFAULT_POLLING_INTERVAL: 5000, // 5 seconds
} as const;

export interface Game {
  id: string;
  title: string;
  description: string;
  start_time: string;
  end_time: string;
  location_id: string;
  host: User;
  players: User[];
  registered_count: number;
  max_players: number;
  skill_level: SkillLevel;
  price: number;
  image_url?: string;
  status: GameStatus;
  created_at: string;
  updated_at: string;
  fixed_game_id?: string;
  location?: Location;
}

export interface GameFilters {
  skillLevel?: SkillLevel;
  date?: string;
  location?: string;
  status?: GameStatus;
}
