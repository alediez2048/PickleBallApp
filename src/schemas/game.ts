import { z } from 'zod';
import { SkillLevel, GameStatus } from '@/types/game';

// Create literal union types from our const maps
const skillLevelSchema = z.enum([
  SkillLevel.Beginner,
  SkillLevel.Intermediate,
  SkillLevel.Advanced,
  SkillLevel.Open,
]);

const gameStatusSchema = z.enum([
  GameStatus.Upcoming,
  GameStatus.InProgress,
  GameStatus.Completed,
  GameStatus.Cancelled,
]);

export const locationSchema = z.object({
  id: z.string().min(1),
  name: z.string().min(1),
  address: z.string().min(1),
  city: z.string().min(1),
  state: z.string().length(2),
  zipCode: z.string().regex(/^\d{5}(-\d{4})?$/),
  coordinates: z.object({
    latitude: z.number().min(-90).max(90),
    longitude: z.number().min(-180).max(180),
  }),
});

export const userSchema = z.object({
  id: z.string().min(1),
  name: z.string().min(2),
  email: z.string().email(),
  skillLevel: skillLevelSchema,
  rating: z.number().min(0).max(5).optional(),
});

export const baseGameSchema = z.object({
  id: z.string().min(1),
  title: z.string().min(3).max(100),
  date: z.date(),
  location: locationSchema,
  maxPlayers: z.number().int().min(2).max(8),
  currentPlayers: z.number().int().min(1),
  skillLevel: skillLevelSchema,
  price: z.number().min(0),
  host: userSchema,
  status: gameStatusSchema,
});

export const gameSchema = baseGameSchema.refine(
  (data) => data.currentPlayers <= data.maxPlayers,
  {
    message: "Current players cannot exceed maximum players",
    path: ["currentPlayers"],
  }
);

// Create type-safe validation functions
export const validateGame = (data: unknown) => gameSchema.safeParse(data);
export const validateLocation = (data: unknown) => locationSchema.safeParse(data);
export const validateUser = (data: unknown) => userSchema.safeParse(data);

// Partial schemas for updates
export const partialGameSchema = z.object({
  id: z.string().min(1).optional(),
  title: z.string().min(3).max(100).optional(),
  date: z.date().optional(),
  location: locationSchema.optional(),
  maxPlayers: z.number().int().min(2).max(8).optional(),
  currentPlayers: z.number().int().min(1).optional(),
  skillLevel: skillLevelSchema.optional(),
  price: z.number().min(0).optional(),
  host: userSchema.optional(),
  status: gameStatusSchema.optional(),
}).refine(
  (data) => !data.currentPlayers || !data.maxPlayers || data.currentPlayers <= data.maxPlayers,
  {
    message: "Current players cannot exceed maximum players",
    path: ["currentPlayers"],
  }
);

export const partialLocationSchema = z.object({
  id: z.string().min(1).optional(),
  name: z.string().min(1).optional(),
  address: z.string().min(1).optional(),
  city: z.string().min(1).optional(),
  state: z.string().length(2).optional(),
  zipCode: z.string().regex(/^\d{5}(-\d{4})?$/).optional(),
  coordinates: z.object({
    latitude: z.number().min(-90).max(90).optional(),
    longitude: z.number().min(-180).max(180).optional(),
  }).optional(),
});

export const partialUserSchema = z.object({
  id: z.string().min(1).optional(),
  name: z.string().min(2).optional(),
  email: z.string().email().optional(),
  skillLevel: skillLevelSchema.optional(),
  rating: z.number().min(0).max(5).optional(),
});

// Create type-safe validation functions for partial data
export const validatePartialGame = (data: unknown) => partialGameSchema.safeParse(data);
export const validatePartialLocation = (data: unknown) => partialLocationSchema.safeParse(data);
export const validatePartialUser = (data: unknown) => partialUserSchema.safeParse(data); 