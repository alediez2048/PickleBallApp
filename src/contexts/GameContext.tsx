import React, { createContext, useContext, useReducer, useCallback, useEffect } from 'react';
import { Game, SkillLevel } from '@/types/game';
import { gamesApi, GameFilters } from '@/services/api/games';
import { ApiError } from '@/services/api/client';

// State interface
interface GameState {
  games: Game[];
  loading: boolean;
  error: string | null;
  optimisticUpdates: Map<string, Game>;
  prefetchedGames: Map<string, Game>;
}

// Action types
type GameAction =
  | { type: 'SET_LOADING'; payload: boolean }
  | { type: 'SET_ERROR'; payload: string | null }
  | { type: 'SET_GAMES'; payload: Game[] }
  | { type: 'ADD_GAME'; payload: Game }
  | { type: 'UPDATE_GAME'; payload: Game }
  | { type: 'DELETE_GAME'; payload: string }
  | { type: 'SET_OPTIMISTIC_UPDATE'; payload: { id: string; game: Game } }
  | { type: 'CLEAR_OPTIMISTIC_UPDATE'; payload: string }
  | { type: 'SET_PREFETCHED_GAME'; payload: { id: string; game: Game } }
  | { type: 'CLEAR_PREFETCHED_GAME'; payload: string };

// Initial state
const initialState: GameState = {
  games: [],
  loading: false,
  error: null,
  optimisticUpdates: new Map(),
  prefetchedGames: new Map(),
};

// Reducer
function gameReducer(state: GameState, action: GameAction): GameState {
  switch (action.type) {
    case 'SET_LOADING':
      return { ...state, loading: action.payload };
    case 'SET_ERROR':
      return { ...state, error: action.payload };
    case 'SET_GAMES':
      return { ...state, games: action.payload };
    case 'ADD_GAME':
      return { ...state, games: [...state.games, action.payload] };
    case 'UPDATE_GAME': {
      const newGames = state.games.map(game => 
        game.id === action.payload.id ? action.payload : game
      );
      const newOptimisticUpdates = new Map(state.optimisticUpdates);
      newOptimisticUpdates.delete(action.payload.id);
      return {
        ...state,
        games: newGames,
        optimisticUpdates: newOptimisticUpdates,
      };
    }
    case 'DELETE_GAME': {
      const newGames = state.games.filter(game => game.id !== action.payload);
      const newOptimisticUpdates = new Map(state.optimisticUpdates);
      newOptimisticUpdates.delete(action.payload);
      return {
        ...state,
        games: newGames,
        optimisticUpdates: newOptimisticUpdates,
      };
    }
    case 'SET_OPTIMISTIC_UPDATE': {
      const newOptimisticUpdates = new Map(state.optimisticUpdates);
      newOptimisticUpdates.set(action.payload.id, action.payload.game);
      return {
        ...state,
        optimisticUpdates: newOptimisticUpdates,
      };
    }
    case 'CLEAR_OPTIMISTIC_UPDATE': {
      const newOptimisticUpdates = new Map(state.optimisticUpdates);
      newOptimisticUpdates.delete(action.payload);
      return {
        ...state,
        optimisticUpdates: newOptimisticUpdates,
      };
    }
    case 'SET_PREFETCHED_GAME': {
      const newPrefetchedGames = new Map(state.prefetchedGames);
      newPrefetchedGames.set(action.payload.id, action.payload.game);
      return {
        ...state,
        prefetchedGames: newPrefetchedGames,
      };
    }
    case 'CLEAR_PREFETCHED_GAME': {
      const newPrefetchedGames = new Map(state.prefetchedGames);
      newPrefetchedGames.delete(action.payload);
      return {
        ...state,
        prefetchedGames: newPrefetchedGames,
      };
    }
    default:
      return state;
  }
}

// Context interface
interface GameContextType extends Omit<GameState, 'optimisticUpdates' | 'prefetchedGames'> {
  fetchGames: (filters?: GameFilters) => Promise<void>;
  createGame: (gameData: Omit<Game, 'id' | 'status'>) => Promise<void>;
  updateGame: (gameId: string, gameData: Partial<Game>) => Promise<void>;
  deleteGame: (gameId: string) => Promise<void>;
  joinGame: (gameId: string) => Promise<void>;
  leaveGame: (gameId: string) => Promise<void>;
  prefetchGame: (gameId: string) => Promise<void>;
  getGame: (gameId: string) => Game | undefined;
}

// Create context
const GameContext = createContext<GameContextType | undefined>(undefined);

// Provider component
export function GameProvider({ children }: { children: React.ReactNode }) {
  const [state, dispatch] = useReducer(gameReducer, initialState);

  const handleError = (error: unknown) => {
    const message = error instanceof ApiError 
      ? error.message 
      : 'An unexpected error occurred';
    dispatch({ type: 'SET_ERROR', payload: message });
  };

  const getGame = useCallback((gameId: string): Game | undefined => {
    // Check optimistic updates first
    const optimisticGame = state.optimisticUpdates.get(gameId);
    if (optimisticGame) return optimisticGame;

    // Then check prefetched games
    const prefetchedGame = state.prefetchedGames.get(gameId);
    if (prefetchedGame) return prefetchedGame;

    // Finally check regular games
    return state.games.find(game => game.id === gameId);
  }, [state.games, state.optimisticUpdates, state.prefetchedGames]);

  const prefetchGame = useCallback(async (gameId: string) => {
    try {
      const response = await gamesApi.getGame(gameId);
      dispatch({ 
        type: 'SET_PREFETCHED_GAME', 
        payload: { id: gameId, game: response.data } 
      });
    } catch (error) {
      console.error('Error prefetching game:', error);
    }
  }, []);

  const fetchGames = useCallback(async (filters?: GameFilters) => {
    dispatch({ type: 'SET_LOADING', payload: true });
    dispatch({ type: 'SET_ERROR', payload: null });
    try {
      const response = await gamesApi.getGames(filters);
      dispatch({ type: 'SET_GAMES', payload: response.data });
    } catch (error) {
      handleError(error);
    } finally {
      dispatch({ type: 'SET_LOADING', payload: false });
    }
  }, []);

  const createGame = useCallback(async (gameData: Omit<Game, 'id' | 'status'>) => {
    dispatch({ type: 'SET_LOADING', payload: true });
    dispatch({ type: 'SET_ERROR', payload: null });
    try {
      const response = await gamesApi.createGame(gameData);
      dispatch({ type: 'ADD_GAME', payload: response.data });
    } catch (error) {
      handleError(error);
      throw error;
    } finally {
      dispatch({ type: 'SET_LOADING', payload: false });
    }
  }, []);

  const updateGame = useCallback(async (gameId: string, gameData: Partial<Game>) => {
    const currentGame = getGame(gameId);
    if (!currentGame) throw new Error('Game not found');

    // Apply optimistic update
    const optimisticGame = { ...currentGame, ...gameData };
    dispatch({ 
      type: 'SET_OPTIMISTIC_UPDATE', 
      payload: { id: gameId, game: optimisticGame } 
    });

    try {
      const response = await gamesApi.updateGame(gameId, gameData);
      dispatch({ type: 'UPDATE_GAME', payload: response.data });
    } catch (error) {
      // Revert optimistic update
      dispatch({ type: 'CLEAR_OPTIMISTIC_UPDATE', payload: gameId });
      handleError(error);
      throw error;
    }
  }, [getGame]);

  const deleteGame = useCallback(async (gameId: string) => {
    // Optimistically remove the game
    const deletedGame = getGame(gameId);
    if (deletedGame) {
      dispatch({ 
        type: 'SET_OPTIMISTIC_UPDATE', 
        payload: { id: gameId, game: { ...deletedGame, status: 'Cancelled' } } 
      });
    }

    try {
      await gamesApi.deleteGame(gameId);
      dispatch({ type: 'DELETE_GAME', payload: gameId });
    } catch (error) {
      // Revert optimistic update
      dispatch({ type: 'CLEAR_OPTIMISTIC_UPDATE', payload: gameId });
      handleError(error);
      throw error;
    }
  }, [getGame]);

  const joinGame = useCallback(async (gameId: string) => {
    const currentGame = getGame(gameId);
    if (!currentGame) throw new Error('Game not found');

    // Apply optimistic update
    const optimisticGame = { 
      ...currentGame, 
      currentPlayers: currentGame.currentPlayers + 1 
    };
    dispatch({ 
      type: 'SET_OPTIMISTIC_UPDATE', 
      payload: { id: gameId, game: optimisticGame } 
    });

    try {
      const response = await gamesApi.joinGame(gameId);
      dispatch({ type: 'UPDATE_GAME', payload: response.data });
    } catch (error) {
      // Revert optimistic update
      dispatch({ type: 'CLEAR_OPTIMISTIC_UPDATE', payload: gameId });
      handleError(error);
      throw error;
    }
  }, [getGame]);

  const leaveGame = useCallback(async (gameId: string) => {
    const currentGame = getGame(gameId);
    if (!currentGame) throw new Error('Game not found');

    // Apply optimistic update
    const optimisticGame = { 
      ...currentGame, 
      currentPlayers: currentGame.currentPlayers - 1 
    };
    dispatch({ 
      type: 'SET_OPTIMISTIC_UPDATE', 
      payload: { id: gameId, game: optimisticGame } 
    });

    try {
      const response = await gamesApi.leaveGame(gameId);
      dispatch({ type: 'UPDATE_GAME', payload: response.data });
    } catch (error) {
      // Revert optimistic update
      dispatch({ type: 'CLEAR_OPTIMISTIC_UPDATE', payload: gameId });
      handleError(error);
      throw error;
    }
  }, [getGame]);

  const value = {
    games: state.games,
    loading: state.loading,
    error: state.error,
    fetchGames,
    createGame,
    updateGame,
    deleteGame,
    joinGame,
    leaveGame,
    prefetchGame,
    getGame,
  };

  return <GameContext.Provider value={value}>{children}</GameContext.Provider>;
}

// Hook for using the context
export function useGames() {
  const context = useContext(GameContext);
  if (context === undefined) {
    throw new Error('useGames must be used within a GameProvider');
  }
  return context;
} 