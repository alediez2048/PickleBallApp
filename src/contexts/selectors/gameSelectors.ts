import { useMemo } from 'react';
import { Game, SkillLevel, GameStatus } from '@/types/games';
import { useGames } from '../GameContext';

// Basic selectors
export const useGameById = (gameId: string) => {
  const { getGame } = useGames();
  return useMemo(() => getGame(gameId), [gameId, getGame]);
};

export const useGamesByStatus = (status: GameStatus) => {
  const { games } = useGames();
  return useMemo(
    () => games.filter(game => game.status === status),
    [games, status]
  );
};

export const useGamesBySkillLevel = (skillLevel: SkillLevel) => {
  const { games } = useGames();
  return useMemo(
    () => games.filter(game => game.skillLevel === skillLevel),
    [games, skillLevel]
  );
};

// Computed selectors
export const useUpcomingGames = () => {
  const { games } = useGames();
  return useMemo(
    () => games.filter(
      game => game.status === GameStatus.Upcoming && new Date(game.date) > new Date()
    ).sort((a, b) => new Date(a.date).getTime() - new Date(b.date).getTime()),
    [games]
  );
};

export const useGameStats = () => {
  const { games } = useGames();
  return useMemo(() => ({
    total: games.length,
    upcoming: games.filter(game => game.status === GameStatus.Upcoming).length,
    inProgress: games.filter(game => game.status === GameStatus.InProgress).length,
    completed: games.filter(game => game.status === GameStatus.Completed).length,
    cancelled: games.filter(game => game.status === GameStatus.Cancelled).length,
    averagePrice: games.reduce((acc, game) => acc + game.price, 0) / games.length || 0,
  }), [games]);
};

// Complex selectors
export const useFilteredGames = (filters: {
  status?: GameStatus;
  skillLevel?: SkillLevel;
  minPrice?: number;
  maxPrice?: number;
  dateRange?: { start: Date; end: Date };
}) => {
  const { games } = useGames();
  
  return useMemo(() => {
    return games.filter(game => {
      if (filters.status && game.status !== filters.status) return false;
      if (filters.skillLevel && game.skillLevel !== filters.skillLevel) return false;
      if (filters.minPrice && game.price < filters.minPrice) return false;
      if (filters.maxPrice && game.price > filters.maxPrice) return false;
      if (filters.dateRange) {
        const gameDate = new Date(game.date);
        if (gameDate < filters.dateRange.start || gameDate > filters.dateRange.end) {
          return false;
        }
      }
      return true;
    });
  }, [games, filters]);
};

// Pagination selector
export const usePaginatedGames = (page: number, pageSize: number) => {
  const { games } = useGames();
  
  return useMemo(() => {
    const start = (page - 1) * pageSize;
    const end = start + pageSize;
    return {
      games: games.slice(start, end),
      totalPages: Math.ceil(games.length / pageSize),
      currentPage: page,
      hasNextPage: end < games.length,
      hasPreviousPage: page > 1,
    };
  }, [games, page, pageSize]);
}; 