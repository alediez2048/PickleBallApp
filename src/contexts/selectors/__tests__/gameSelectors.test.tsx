import React from 'react';
import renderer from 'react-test-renderer';
import { GameProvider } from '../../GameContext';
import * as gameSelectors from '../gameSelectors';
import { Game, GameStatus, SkillLevel } from '@/types/game';

// Mock data
const mockGames: Game[] = [
  {
    id: '1',
    title: 'Game 1',
    description: 'Test game description',
    startTime: '2024-03-25T10:00:00',
    endTime: '2024-03-25T12:00:00',
    location: {
      id: '1',
      name: 'Court 1',
      address: '123 Main St',
      city: 'Test City',
      state: 'TS',
      zipCode: '12345',
      coordinates: { latitude: 0, longitude: 0 },
    },
    maxPlayers: 4,
    registeredCount: 2,
    players: [],
    skillLevel: SkillLevel.Beginner,
    price: 10,
    host: {
      id: '1',
      name: 'Host 1',
      email: 'host1@test.com',
      skillLevel: SkillLevel.Intermediate,
    },
    status: GameStatus.Upcoming,
    createdAt: '2024-03-20T10:00:00',
    updatedAt: '2024-03-20T10:00:00',
  },
];

// Mock GameContext
jest.mock('../../GameContext', () => ({
  GameProvider: ({ children }: { children: React.ReactNode }) => children,
  useGames: () => ({
    games: mockGames,
    getGame: (id: string) => mockGames.find(game => game.id === id),
    isLoading: false,
    loadMore: jest.fn(),
  }),
}));

// Components to test the hooks
interface GameByIdProps {
  id: string;
}

const GameByIdComponent: React.FC<GameByIdProps> = ({ id }) => {
  const game = gameSelectors.useGameById(id);
  return <div data-testid="game-by-id">{game ? game.title : 'Not found'}</div>;
};

interface GamesByStatusProps {
  status: GameStatus;
}

const GamesByStatusComponent: React.FC<GamesByStatusProps> = ({ status }) => {
  const games = gameSelectors.useGamesByStatus(status);
  return (
    <div data-testid="games-by-status">
      {games.map(game => <div key={game.id}>{game.title}</div>)}
    </div>
  );
};

interface GamesBySkillLevelProps {
  skillLevel: SkillLevel;
}

const GamesBySkillLevelComponent: React.FC<GamesBySkillLevelProps> = ({ skillLevel }) => {
  const games = gameSelectors.useGamesBySkillLevel(skillLevel);
  return (
    <div data-testid="games-by-skill">
      {games.map(game => <div key={game.id}>{game.title}</div>)}
    </div>
  );
};

const GameStatsComponent: React.FC = () => {
  const stats = gameSelectors.useGameStats();
  return (
    <div data-testid="game-stats">
      <div>Total: {stats.total}</div>
      <div>Upcoming: {stats.upcoming}</div>
      <div>In Progress: {stats.inProgress}</div>
      <div>Completed: {stats.completed}</div>
      <div>Cancelled: {stats.cancelled}</div>
      <div>Average Price: {stats.averagePrice}</div>
    </div>
  );
};

interface FilteredGamesProps {
  filters: {
    status?: GameStatus;
    skillLevel?: SkillLevel;
    minPrice?: number;
    maxPrice?: number;
    dateRange?: { start: Date; end: Date };
  };
}

const FilteredGamesComponent: React.FC<FilteredGamesProps> = ({ filters }) => {
  const games = gameSelectors.useFilteredGames(filters);
  return (
    <div data-testid="filtered-games">
      {games.map(game => <div key={game.id}>{game.title}</div>)}
    </div>
  );
};

interface PaginatedGamesProps {
  page: number;
  pageSize: number;
}

const PaginatedGamesComponent: React.FC<PaginatedGamesProps> = ({ page, pageSize }) => {
  const pagination = gameSelectors.usePaginatedGames(page, pageSize);
  return (
    <div data-testid="paginated-games">
      <div>Page: {pagination.currentPage}</div>
      <div>Total Pages: {pagination.totalPages}</div>
      <div>
        {pagination.games.map(game => <div key={game.id}>{game.title}</div>)}
      </div>
    </div>
  );
};

describe('Game Selectors', () => {
  describe('useGameById', () => {
    it('returns the correct game by id', () => {
      const tree = renderer.create(<GameByIdComponent id="1" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('returns undefined for non-existent game', () => {
      const tree = renderer.create(<GameByIdComponent id="999" />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('useGamesByStatus', () => {
    it('filters games by status', () => {
      const tree = renderer.create(<GamesByStatusComponent status={GameStatus.Upcoming} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('useGamesBySkillLevel', () => {
    it('filters games by skill level', () => {
      const tree = renderer.create(<GamesBySkillLevelComponent skillLevel={SkillLevel.Beginner} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('useGameStats', () => {
    it('calculates correct game statistics', () => {
      const tree = renderer.create(<GameStatsComponent />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('useFilteredGames', () => {
    it('filters games based on multiple criteria', () => {
      const filters = {
        status: GameStatus.Upcoming,
        skillLevel: SkillLevel.Beginner,
        minPrice: 5,
        maxPrice: 15,
      };
      const tree = renderer.create(<FilteredGamesComponent filters={filters} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('returns empty array when no games match filters', () => {
      const filters = {
        minPrice: 100,
      };
      const tree = renderer.create(<FilteredGamesComponent filters={filters} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('usePaginatedGames', () => {
    it('returns correct pagination data', () => {
      const tree = renderer.create(<PaginatedGamesComponent page={1} pageSize={10} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
}); 