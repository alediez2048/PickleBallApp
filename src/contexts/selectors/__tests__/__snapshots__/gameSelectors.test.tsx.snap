// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Selectors useFilteredGames filters games based on multiple criteria 1`] = `null`;

exports[`Game Selectors useFilteredGames returns empty array when no games match filters 1`] = `null`;

exports[`Game Selectors useGameById returns the correct game by id 1`] = `null`;

exports[`Game Selectors useGameById returns undefined for non-existent game 1`] = `null`;

exports[`Game Selectors useGameStats calculates correct game statistics 1`] = `null`;

exports[`Game Selectors useGamesBySkillLevel filters games by skill level 1`] = `null`;

exports[`Game Selectors useGamesByStatus filters games by status 1`] = `null`;

exports[`Game Selectors usePaginatedGames returns correct pagination data 1`] = `null`;
